libdnf5/base/goal.cpp

```diff
@@ -2206,6 +2412,130 @@ GoalProblem Goal::Impl::add_reason_change_to_goal(
+        transaction::GroupReplay group_replay;
+        group_replay.group_id = group.to_string();
+        // Do not revert UPGRADE for groups. Groups don't have an upgrade path so they cannot be
+        // upgraded or downgraded. The UPGRADE action is basically a synchronization with
+        // current group definition. Reverting synchronization is again synchronization but
+        // with the older group definition, this happens automatically by reverting the rpm
+        // actions. We only have to keep the UPGRADE action to have a record of the operation.
+        if (group.get_action() != transaction::TransactionItemAction::UPGRADE) {
+            auto reverted_action = REVERT_ACTION.find(group.get_action());
+            if (reverted_action == REVERT_ACTION.end()) {
+                libdnf_throw_assertion(
+                    "Cannot revert action: \"{}\"", transaction_item_action_to_string(group.get_action()));
+            }
+            group_replay.action = reverted_action->second;
+        } else {
+            group_replay.action = transaction::TransactionItemAction::UPGRADE;
+        }
+
+        if (group_replay.action == Action::INSTALL && group.get_reason() == Reason::CLEAN) {
+            group_replay.reason = Reason::DEPENDENCY;
+        } else if (group_replay.action == Action::REMOVE && group.get_reason() == Reason::DEPENDENCY) {
+            group_replay.reason = Reason::CLEAN;
+        } else {
+            group_replay.reason = group.get_reason();
+        }
+
+        // Don't fill in @System repo_id because it's not possible to perform forward actions from it
+        if (group.get_repoid() != "@System") {
+            group_replay.repo_id = group.get_repoid();
+        }
+
+        replay.groups.push_back(group_replay);
+    }
+
+    for (const auto & env : reverting_transaction.get_comps_environments()) {
+        transaction::EnvironmentReplay env_replay;
+        env_replay.environment_id = env.to_string();
+        // Do not revert UPGRADE for environments. Environments don't have an upgrade path so they cannot be
+        // upgraded or downgraded. The UPGRADE action is basically a synchronization with
+        // current environment definition. Reverting synchronization is again synchronization but
+        // with the older environment definition, this happens automatically by reverting the rpm
+        // actions. We only have to keep the UPGRADE action to have a record of the operation.
+        if (env.get_action() != transaction::TransactionItemAction::UPGRADE) {
+            auto reverted_action = REVERT_ACTION.find(env.get_action());
+            if (reverted_action == REVERT_ACTION.end()) {
+                libdnf_throw_assertion(
+                    "Cannot revert action: \"{}\"", transaction_item_action_to_string(env.get_action()));
+            }
+            env_replay.action = reverted_action->second;
+        } else {
+            env_replay.action = transaction::TransactionItemAction::UPGRADE;
```

Comment:
I am recommending to not present the operation as an upgrade, because in background it does nothing. But if new group appeared then it might modify groups.

Reply:
Ok, I have turned into it a log event. Now a revert of a revert of group upgrade won't mention the group at all because the group action will disappear on the first revert but I think that is acceptable.